eth_account/_utils/encode_typed_data/encoding_and_hashing.py

```diff
@@ -0,0 +1,221 @@
+    custom_types = set(types.keys())
+    custom_types_that_are_deps = set()
+
+    for type in custom_types:
+        type_fields = types[type]
+        for field in type_fields:
+            parsed_type = parse_core_array_type(field["type"])
+            if parsed_type in custom_types and parsed_type != type:
+                custom_types_that_are_deps.add(parsed_type)
+
+    primary_type = list(custom_types.difference(custom_types_that_are_deps))
+    if len(primary_type) == 1:
+        return primary_type[0]
+    else:
+        raise ValueError("Unable to determine primary type")
+
+
+def encode_field(
+    types: Dict[str, List[Dict[str, str]]],
+    name: str,
+    type_: str,
+    value: Any,
+) -> Tuple[str, Union[int, bytes]]:
+    if type_ in types.keys():
+        # type is a custom type
+        if value is None:
+            return (
+                "bytes32",
+                b"\x00" * 32,
+            )
+        else:
+            return (
+                "bytes32",
+                keccak(encode_data(type_, types, value)),
+            )
+
+    if type_ in ["string", "bytes"] and value is None:
+        return (
+            "bytes32",
+            b"",
+        )
+
+    # None is allowed only for custom and dynamic types
+    if value is None:
+        raise ValueError(f"missing value for field {name} of type {type_}")
+
+    if is_array_type(type_):
+        type_ = parse_parent_array_type(type_)
+        type_value_pairs = [encode_field(types, name, type_, item) for item in value]
+        data_types, data_hashes = zip(*type_value_pairs)
+        return ("bytes32", keccak(encode(data_types, data_hashes)))
+
+    if type_ == "bool":
+        return (type_, coerce_bool(value))
+
+    if type_ == "bytes":
+        if is_hexstr(value):
+            value = to_bytes(hexstr=value)
+        else:
+            value = to_bytes(value)
+        return ("bytes32", keccak(value))
+
+    if type_ == "string":
+        if isinstance(value, int):
+            value = to_bytes(value)
+        else:
+            value = to_bytes(text=value)
+        return ("bytes32", keccak(value))
+
+    # allow string values for int and uint types
+    if type(value) == str and type_.startswith(("int", "uint")):
+        if value[:2] == "0x":
+            return (type_, to_int(hexstr=value))
+        else:
+            return (type_, to_int(text=value))
+
+    return (type_, value)
+
+
+def find_type_dependencies(type_, types, results=None):
+    if results is None:
+        results = set()
+
+    # a type must be a string
+    if not isinstance(type_, str):
+        raise ValueError(
+            "Invalid find_type_dependencies input: expected string, got "
+            f"{type_} of type {type(type_)}"
+        )
+    # get core type if it's an array type
+    type_ = parse_core_array_type(type_)
+
+    # don't look for dependencies of solidity types
+    if type_ in EIP712_SOLIDITY_TYPES:
```

Comment:
tiny nit: two bits of logic with the same return statement can be combined. 

`if...elif` can be used here as well.

```py
if (
    # don't look for dependencies of solidity types
    type_ in EIP712_SOLIDITY_TYPES
    # found a type that's already been added
    or type_ in results
):
    return results